File: python/modules/render/render_signal_analyzer.py
"""Render signal analyzer — verbatim port of AEC3
docs/aec3_extracts/src/aec3/render_signal_analyzer.{cc,h}.

Used by the refined / coarse filter gain compute to:

1. Detect narrow-band (tonal) render regions and zero the corresponding mu
   bins via ``mask_regions_around_narrow_bands(mu)`` (AEC3 cc:143-159).
2. Report ``poor_signal_excitation()`` when any narrow-band counter has been
   sustained for > 10 frames — used to freeze the W update entirely
   (AEC3 cc:96-99 → refined_filter_update_gain.cc startup gate).
3. Report ``narrow_peak_band()`` for diagnostic / further use.

Float-scale conversion: peak detection uses ratios (unitless) so no constant
scaling. The "max_abs > 100" threshold in the strong-narrow-band path is
int16 amplitude; converted to float via /32768 = 3.05e-3.

Configurable for any FFT size; the AEC3 hard-coded kFftLengthBy2 = 64 becomes
runtime ``n_freqs - 1`` (where ``n_freqs = fft_size/2 + 1``).
"""
from typing import Optional

import numpy as np


_COUNTER_THRESHOLD = 5  # AEC3 kCounterThreshold (mask fires when > 5 frames)
_POOR_EXCITATION_THRESHOLD = 10  # AEC3 narrow_band_counter > 10 → poor_excitation
_MAX_ABS_THRESHOLD_INT16 = 100.0  # AEC3 strong-narrow-band amplitude floor
_MAX_ABS_THRESHOLD_FLOAT = _MAX_ABS_THRESHOLD_INT16 / 32768.0   # 3.05e-3
_PEAK_RATIO_NARROW = 3.0           # AEC3 X²[k] > 3 × max(neighbors)
_PEAK_RATIO_STRONG = 100.0         # AEC3 strong-peak: peak_level > 100 × non_peak


class RenderSignalAnalyzer:
    """Per-hop render-signal analyser. Track narrow-band tonal regions in the
    delay-aligned render spectrum + detect dominant narrow peaks.

    Inputs (per call to ``update``):
        render_psd     : per-bin power spectrum |X|² of the (delay-aligned)
                         render block. Length = n_freqs = fft_size/2 + 1.
                         May be None to signal "no delay estimate yet" →
                         counters reset.
        render_block   : optional time-domain render samples for the strong-
                         narrow-band detector's max_abs check. May be None
                         to skip strong-peak detection.

    Outputs (read via methods):
        poor_signal_excitation() -> bool
        narrow_peak_band()       -> Optional[int]
        mask_regions_around_narrow_bands(mu)  -> in-place mu zeroing
    """

    def __init__(self, n_freqs: int, strong_peak_freeze_duration: int = 6) -> None:
        # AEC3 array has size kFftLengthBy2 - 1 (= n_freqs - 2). Indexed by k-1
        # for bins k in [1, kFftLengthBy2 - 1).
        self._n_freqs = int(n_freqs)
        self._n_freqs_by2 = self._n_freqs - 1  # kFftLengthBy2 in AEC3
        if self._n_freqs_by2 < 3:
            raise ValueError(f"n_freqs={n_freqs} too small for narrow-band detect")
        # Per-bin counter, length kFftLengthBy2 - 1 (bins 1..n_freqs-2).
        self._counters = np.zeros(self._n_freqs_by2 - 1, dtype=np.int64)
        # Strong-peak state.
        self._strong_peak_freeze_duration = int(strong_peak_freeze_duration)
        self._narrow_peak_band: Optional[int] = None
        self._narrow_peak_counter = 0

    # -------------------------------------------------------- AEC3 detectors
    def _identify_small_narrow_band_regions(
        self, render_psd: Optional[np.ndarray]
    ) -> None:
        """Mirrors IdentifySmallNarrowBandRegions (cc:33-59).
        Single-channel: channel_counters trivially = 0 or 1 per bin."""
        if render_psd is None:
            # AEC3: no delay → zero all counters.
            self._counters.fill(0)
            return
        X2 = render_psd
        # For k in [1, kFftLengthBy2): check X2[k] > 3 * max(X2[k-1], X2[k+1])
        # We iterate bins 1..n_freqs_by2-1 (inclusive), array index k-1.
        k_lo, k_hi = 1, self._n_freqs_by2  # exclusive upper
        # Vectorise: build view of [k_lo..k_hi), check neighbour condition.
        x_center = X2[k_lo:k_hi]
        x_left = X2[k_lo - 1:k_hi - 1]
        x_right = X2[k_lo + 1:k_hi + 1]
        is_peak = x_center > _PEAK_RATIO_NARROW * np.maximum(x_left, x_right)
        # Update counters: if peak this frame → +1, else reset to 0.
        # (channel_counters > 0 ? counter+1 : 0 — for single channel, equivalent.)
        self._counters = np.where(is_peak, self._counters + 1, 0).astype(np.int64)

    def _identify_strong_narrow_band_component(
        self,
        render_psd: Optional[np.ndarray],
        render_block: Optional[np.ndarray],
    ) -> None:
        """Mirrors IdentifyStrongNarrowBandComponent (cc:62-121)."""
        # Tick down freeze counter — if expired, clear narrow_peak_band.
        if self._narrow_peak_band is not None:
            self._narrow_peak_counter += 1
            if self._narrow_peak_counter > self._strong_peak_freeze_duration:
                self._narrow_peak_band = None
                self._narrow_peak_counter = 0
        if render_psd is None or render_block is None:
            return
        X2 = render_psd
        peak_bin = int(np.argmax(X2))
        # Non-peak power in two windows around peak (avoiding ±5 bins).
        lo_lo = max(0, peak_bin - 14)
        lo_hi = max(lo_lo, peak_bin - 4)
        hi_lo = min(self._n_freqs, peak_bin + 5)
        hi_hi = min(self._n_freqs, peak_bin + 15)
        non_peak_power = 0.0
        if lo_hi > lo_lo:
            non_peak_power = max(non_peak_power, float(np.max(X2[lo_lo:lo_hi])))
        if hi_hi > hi_lo:
            non_peak_power = max(non_peak_power, float(np.max(X2[hi_lo:hi_hi])))
        # max_abs of render time block (float scale).
        max_abs = float(np.max(np.abs(render_block)))
        peak_level = float(X2[peak_bin])
        if (peak_bin > 0
                and max_abs > _MAX_ABS_THRESHOLD_FLOAT
                and peak_level > _PEAK_RATIO_STRONG * non_peak_power):
            self._narrow_peak_band = peak_bin
            self._narrow_peak_counter = 0

    # --------------------------------------------------------------- public
    def update(
        self,
        render_psd: Optional[np.ndarray],
        render_block: Optional[np.ndarray] = None,
    ) -> None:
        """Per-hop update.

        ``render_psd`` should be the delay-aligned partition's |X|² (passing
        the current-frame X² when no delay is locked is acceptable, mirrors
        the AEC3 ``delay_partitions=None`` case → counters reset).
        ``render_block`` is the time-domain render hop (for strong-peak's
        max_abs gate). Both may be None during cold-start.
        """
        self._identify_small_narrow_band_regions(render_psd)
        self._identify_strong_narrow_band_component(render_psd, render_block)

    def poor_signal_excitation(self) -> bool:
        """True iff ANY narrow-band counter > 10 (AEC3 cc:40-45)."""
        if self._counters.size == 0:
            return False
        return bool(np.any(self._counters > _POOR_EXCITATION_THRESHOLD))

    def narrow_peak_band(self) -> Optional[int]:
        return self._narrow_peak_band

    def mask_regions_around_narrow_bands(self, mu: np.ndarray) -> np.ndarray:
        """In-place mask: zero mu for ±2 bins around any peak with counter > 5
        (AEC3 cc:143-159). Returns the same array for chaining convenience.

        Bin edges are special-cased to mirror AEC3 (cc:148-149, 156-157):
          - counter[0]      > threshold  →  zero mu[0], mu[1]
          - counter[k-1]    > threshold  →  zero mu[k-2..k+2] (k in [2, N-1))
          - counter[N-2]    > threshold  →  zero mu[N-1], mu[N]   where N = n_freqs_by2
        """
        if mu.shape[0] != self._n_freqs:
            raise ValueError(
                f"mu has {mu.shape[0]} bins, expected n_freqs={self._n_freqs}"
            )
        N = self._n_freqs_by2  # = n_freqs - 1
        # Edge: counter[0] (bin k=1) → mask bins 0, 1.
        if self._counters[0] > _COUNTER_THRESHOLD:
            mu[0] = 0.0
            mu[1] = 0.0
        # Interior: counter[k-1] (bin k, k in [2, N-1)) → mask bins k-2..k+2.
        for k in range(2, N - 1):
            if self._counters[k - 1] > _COUNTER_THRESHOLD:
                mu[k - 2:k + 3] = 0.0
        # Edge: counter[N-2] (bin k=N-1) → mask bins N-1, N (=n_freqs-1).
        if self._counters[N - 2] > _COUNTER_THRESHOLD:
            mu[N - 1] = 0.0
            mu[N] = 0.0
        return mu

    # ----------------------------------------------------------------- reset
    def reset(self) -> None:
        self._counters.fill(0)
        self._narrow_peak_band = None
        self._narrow_peak_counter = 0

File: c_impl/src/render_signal_analyzer.c
/* render_signal_analyzer.c — byte-equal C port of
 * python/modules/render/render_signal_analyzer.py. See header for parity notes.
 *
 * Constants (mirror the Python module-level constants):
 *   _COUNTER_THRESHOLD        = 5    (mask fires when counter > 5)
 *   _POOR_EXCITATION_THRESHOLD= 10   (counter > 10 → poor_excitation)
 *   _MAX_ABS_THRESHOLD_FLOAT  = 100.0 / 32768.0  (strong-peak amplitude floor)
 *   _PEAK_RATIO_NARROW        = 3.0  (small narrow band)
 *   _PEAK_RATIO_STRONG        = 100.0 (strong peak)
 */
#include "render_signal_analyzer.h"

#include <math.h>
#include <stddef.h>
#include <string.h>

#define RSA_COUNTER_THRESHOLD          5
#define RSA_POOR_EXCITATION_THRESHOLD  10
#define RSA_MAX_ABS_THRESHOLD_FLOAT    (100.0f / 32768.0f)   /* 3.0517578125e-3 */
#define RSA_PEAK_RATIO_NARROW          3.0f
#define RSA_PEAK_RATIO_STRONG          100.0f

void rsa_init(RenderSignalAnalyzer *m, int64_t *counters_storage, int n_freqs,
              int strong_peak_freeze_duration) {
    m->n_freqs = n_freqs;
    m->n_freqs_by2 = n_freqs - 1;          /* kFftLengthBy2 */
    m->n_counters = m->n_freqs_by2 - 1;    /* = n_freqs - 2 */
    m->counters = counters_storage;
    m->strong_peak_freeze_duration = strong_peak_freeze_duration;
    m->narrow_peak_band = -1;
    m->narrow_peak_counter = 0;
    memset(m->counters, 0, (size_t)m->n_counters * sizeof(int64_t));
}

void rsa_reset(RenderSignalAnalyzer *m) {
    memset(m->counters, 0, (size_t)m->n_counters * sizeof(int64_t));
    m->narrow_peak_band = -1;
    m->narrow_peak_counter = 0;
}

/* Mirrors _identify_small_narrow_band_regions (cc:33-59). render_psd NULL →
 * zero all counters. Else per bin k in [1, n_freqs_by2):
 *   is_peak = X2[k] > 3.0f * fmaxf(X2[k-1], X2[k+1])   (ALL float32)
 *   counter[k-1] = is_peak ? counter[k-1] + 1 : 0
 * Counter array index runs 0..n_counters-1 for k = 1..n_freqs_by2-1. */
static void identify_small_narrow_band_regions(RenderSignalAnalyzer *m,
                                               const float *render_psd) {
    int k;
    if (render_psd == NULL) {
        memset(m->counters, 0, (size_t)m->n_counters * sizeof(int64_t));
        return;
    }
    /* k_lo=1, k_hi=n_freqs_by2 (exclusive). Index into counters is k-1. */
    for (k = 1; k < m->n_freqs_by2; ++k) {
        float x_center = render_psd[k];
        float x_left   = render_psd[k - 1];
        float x_right  = render_psd[k + 1];
        float neigh    = x_left > x_right ? x_left : x_right; /* np.maximum */
        float thresh   = RSA_PEAK_RATIO_NARROW * neigh; /* f32*f32 -> f32 */
        int idx = k - 1;
        if (x_center > thresh) {
            m->counters[idx] = m->counters[idx] + 1;
        } else {
            m->counters[idx] = 0;
        }
    }
}

/* Mirrors _identify_strong_narrow_band_component (cc:62-121). */
static void identify_strong_narrow_band_component(RenderSignalAnalyzer *m,
                                                  const float *render_psd,
                                                  const float *render_block,
                                                  int render_block_len) {
    int peak_bin, lo_lo, lo_hi, hi_lo, hi_hi, i;
    float non_peak_power, max_abs, peak_level;
    float wmax;

    /* Tick down freeze counter. */
    if (m->narrow_peak_band != -1) {
        m->narrow_peak_counter += 1;
        if (m->narrow_peak_counter > m->strong_peak_freeze_duration) {
            m->narrow_peak_band = -1;
            m->narrow_peak_counter = 0;
        }
    }
    if (render_psd == NULL || render_block == NULL) {
        return;
    }

    /* peak_bin = int(np.argmax(X2)) — first index of the maximum. */
    peak_bin = 0;
    for (i = 1; i < m->n_freqs; ++i) {
        if (render_psd[i] > render_psd[peak_bin]) {
            peak_bin = i;
        }
    }

    /* Non-peak power windows (avoiding ±5 bins around peak). */
    lo_lo = peak_bin - 14; if (lo_lo < 0) lo_lo = 0;
    lo_hi = peak_bin - 4;  if (lo_hi < lo_lo) lo_hi = lo_lo;
    hi_lo = peak_bin + 5;  if (hi_lo > m->n_freqs) hi_lo = m->n_freqs;
    hi_hi = peak_bin + 15; if (hi_hi > m->n_freqs) hi_hi = m->n_freqs;

    non_peak_power = 0.0f;
    if (lo_hi > lo_lo) {
        wmax = render_psd[lo_lo];               /* np.max over float32 window */
        for (i = lo_lo + 1; i < lo_hi; ++i) {
            if (render_psd[i] > wmax) wmax = render_psd[i];
        }
        if (wmax > non_peak_power) non_peak_power = wmax;
    }
    if (hi_hi > hi_lo) {
        wmax = render_psd[hi_lo];
        for (i = hi_lo + 1; i < hi_hi; ++i) {
            if (render_psd[i] > wmax) wmax = render_psd[i];
        }
        if (wmax > non_peak_power) non_peak_power = wmax;
    }

    /* max_abs of render time block: np.max(np.abs(render_block)), kept in
     * float32 (float32-by-design; formerly widened via float() to f64). */
    {
        float amax = (render_block_len > 0) ? fabsf(render_block[0]) : 0.0f;
        for (i = 1; i < render_block_len; ++i) {
            float a = fabsf(render_block[i]);
            if (a > amax) amax = a;
        }
        max_abs = amax;
    }

    peak_level = render_psd[peak_bin];   /* X2[peak_bin], float32-by-design */

    if (peak_bin > 0
            && max_abs > RSA_MAX_ABS_THRESHOLD_FLOAT
            && peak_level > RSA_PEAK_RATIO_STRONG * non_peak_power) {
        m->narrow_peak_band = peak_bin;
        m->narrow_peak_counter = 0;
    }
}

void rsa_update(RenderSignalAnalyzer *m, const float *render_psd,
                const float *render_block, int render_block_len) {
    identify_small_narrow_band_regions(m, render_psd);
    identify_strong_narrow_band_component(m, render_psd, render_block,
                                          render_block_len);
}

int rsa_poor_signal_excitation(const RenderSignalAnalyzer *m) {
    int i;
    if (m->n_counters == 0) return 0;
    for (i = 0; i < m->n_counters; ++i) {
        if (m->counters[i] > RSA_POOR_EXCITATION_THRESHOLD) return 1;
    }
    return 0;
}

int rsa_narrow_peak_band(const RenderSignalAnalyzer *m) {
    return m->narrow_peak_band;
}

/* In-place mask: zero mu for ±2 bins around any peak with counter > 5.
 *   counter[0]   > thr → mu[0], mu[1]
 *   counter[k-1] > thr → mu[k-2..k+2]   (k in [2, N-1))
 *   counter[N-2] > thr → mu[N-1], mu[N]     where N = n_freqs_by2
 */
void rsa_mask_regions_around_narrow_bands(const RenderSignalAnalyzer *m, float *mu) {
    int N = m->n_freqs_by2;  /* = n_freqs - 1 */
    int k, j;

    if (m->counters[0] > RSA_COUNTER_THRESHOLD) {
        mu[0] = 0.0f;
        mu[1] = 0.0f;
    }
    for (k = 2; k < N - 1; ++k) {
        if (m->counters[k - 1] > RSA_COUNTER_THRESHOLD) {
            /* mu[k-2 : k+3] = 0.0  (k+2 inclusive, clamped to n_freqs-1) */
            for (j = k - 2; j <= k + 2; ++j) {
                if (j >= 0 && j < m->n_freqs) mu[j] = 0.0f;
            }
        }
    }
    if (m->counters[N - 2] > RSA_COUNTER_THRESHOLD) {
        mu[N - 1] = 0.0f;
        mu[N] = 0.0f;
    }
}
